Backport #19195 to selenium-webdriver/v2

jablko opened the pull request:
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- ~[ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.~
- ~[ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.~

jessetrinity commented:
I think you forgot to add the link, which would help get some reviewer interest.

sandersn commented:
@jablko this seems like a good change, but I don't want to change such a popular package without some reviewers.

@BillArmstrong @Kuniwak @cnishina @SupernaviX @bendxn @oddui @beta-vulgaris @Dzenly @thislooksfun  can you take a look? This looks like a straightforward port of #19195 so I think it's fine.

oddui commented:
Looks good to me. However I don't use v2 so didn't test it in a project.

LukasHeimann commented:
Hi everyone,
this PR causes some issues. I'm using @ui5/uiveri5, which depends on v2 of selenium webdriver and my tsc builds just broke.

Minimal way to show this:

1. `npm init -y`
2. `npm i typescript @types/selenium-webdriver`
3. `npx tsc --init`
4. Create some typescript file
5. `npx tsc`

```
node_modules/@types/selenium-webdriver/index.d.ts:805:13 - error TS2430: Interface 'IThenable<T>' incorrectly extends interface 'PromiseLike<T>'.
  The types of 'then(...).then' are incompatible between these types.
    Type '<R>(opt_callback?: ((value: TResult1) => R | IThenable<R>) | undefined, opt_errback?: ((error: any) => any) | undefined) => Promise<R>' is not assignable to type '<TResult1 = TResult1 | TResult2, TResult2 = never>(onfulfilled?: ((value: TResult1 | TResult2) => TResult1 | PromiseLike<TResult1>) | null | undefined, onrejected?: ((reason: any) => TResult2 | PromiseLike<...>) | ... 1 more ... | undefined) => PromiseLike<...>'.
      Types of parameters 'opt_callback' and 'onfulfilled' are incompatible.
        Type '((value: TResult1 | TResult2) => TResult1 | PromiseLike<TResult1>) | null | undefined' is not assignable to type '((value: TResult1) => TResult1 | IThenable<TResult1>) | undefined'.
          Type 'null' is not assignable to type '((value: TResult1) => TResult1 | IThenable<TResult1>) | undefined'.

805   interface IThenable<T> extends PromiseLike<T> {
                ~~~~~~~~~

node_modules/@types/selenium-webdriver/index.d.ts:1038:5 - error TS2416: Property 'then' in type 'Promise<T>' is not assignable to the same property in base type 'IThenable<T>'.
  Types of parameters 'opt_callback' and 'opt_callback' are incompatible.
    Type 'R | PromiseLike<R>' is not assignable to type 'R | IThenable<R>'.
      Type 'PromiseLike<R>' is not assignable to type 'R | IThenable<R>'.
        Type 'PromiseLike<R>' is missing the following properties from type 'IThenable<R>': cancel, isPending, catch

1038     then<R>(opt_callback?: (value: T) => IThenable<R> | R, opt_errback?: (error: any) => any): Promise<R>;
         ~~~~

node_modules/@types/selenium-webdriver/index.d.ts:4835:3 - error TS2416: Property 'serialize' in type 'WebElement' is not assignable to the same property in base type 'Serializable<IWebElementId>'.
  Type '() => Promise<IWebElementId>' is not assignable to type '() => IWebElementId | IThenable<IWebElementId>'.
    Type 'Promise<IWebElementId>' is not assignable to type 'IWebElementId | IThenable<IWebElementId>'.
      Type 'Promise<IWebElementId>' is not assignable to type 'IWebElementId'.
        Index signature is missing in type 'Promise<IWebElementId>'.

4835   serialize(): promise.Promise<IWebElementId>;
       ~~~~~~~~~


Found 3 errors.
```

It works fine when i manually downgrade to the previous version of the types.

@sandersn: Can this change be rolled back? Do you need an additional GH Issue for this?

Thank you very much for your help!

jablko commented:
@LukasHeimann Thank you for the reproducer, I expect this is fixed by de6d31e99ccae27d2dc58982c57c4f2f77b699f7 however I'm still trying to reproduce the error, to confirm that it's fixed.

jablko commented:
I'm confused: The only way I can reproduce this is with `tsc --strictFunctionTypes --strictNullChecks` however I can't get the tests to pass with `strictFunctionTypes` and `strictNullChecks`, even pre this change. Still investigating.

jablko commented:
@LukasHeimann I think #42319 fixes the error you found: I switched on `strictFunctionTypes` and `strictNullChecks` and fixed the errors they report. Thank you for reporting this, I'm sorry for not catching it myself!

LukasHeimann commented:
@jablko Thank you for providing this quick fix! It seems to work now with the default options created by tsc --init.
